From 6321e8153afa1fe541e558224e1daa919fdfe4ca Mon Sep 17 00:00:00 2001
From: Amelia Crate <amelia.crate@chainguard.dev>
Date: Thu, 11 Sep 2025 15:38:43 -0500
Subject: [PATCH] crypto: skip test cases with short input when fips_enabled is
 true

These don't pass because hmac rejects short input in fips mode.
---
 crypto/hkdf.c | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/crypto/hkdf.c b/crypto/hkdf.c
index 82d1b32ca6ce..ac3ce98096c6 100644
--- a/crypto/hkdf.c
+++ b/crypto/hkdf.c
@@ -11,6 +11,7 @@
 #include <crypto/sha2.h>
 #include <crypto/hkdf.h>
 #include <linux/module.h>
+#include <linux/fips.h>
 
 /*
  * HKDF consists of two steps:
@@ -141,6 +142,7 @@ struct hkdf_testvec {
 	u16 info_size;
 	u16 prk_size;
 	u16 okm_size;
+	u8 fips_skip;
 };
 
 /*
@@ -166,6 +168,7 @@ static const struct hkdf_testvec hkdf_sha256_tv[] = {
 			"\x2d\x2d\x0a\x90\xcf\x1a\x5a\x4c\x5d\xb0\x2d\x56\xec\xc4\xc5\xbf"
 			"\x34\x00\x72\x08\xd5\xb8\x87\x18\x58\x65",
 		.okm_size = 42,
+		.fips_skip = 1,
 	}, {
 		.test = "hkdf test with long input",
 		.ikm  = "\x00\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f"
@@ -196,6 +199,7 @@ static const struct hkdf_testvec hkdf_sha256_tv[] = {
 			"\xcc\x30\xc5\x81\x79\xec\x3e\x87\xc1\x4c\x01\xd5\xc1\xf3\x43\x4f"
 			"\x1d\x87",
 		.okm_size = 82,
+		.fips_skip = 0,
 	}, {
 		.test = "hkdf test with zero salt and info",
 		.ikm  = "\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b"
@@ -212,6 +216,7 @@ static const struct hkdf_testvec hkdf_sha256_tv[] = {
 			"\xb8\xa1\x1f\x5c\x5e\xe1\x87\x9e\xc3\x45\x4e\x5f\x3c\x73\x8d\x2d"
 			"\x9d\x20\x13\x95\xfa\xa4\xb6\x1a\x96\xc8",
 		.okm_size = 42,
+		.fips_skip = 1,
 	}, {
 		.test = "hkdf test with short input",
 		.ikm  = "\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b",
@@ -227,6 +232,7 @@ static const struct hkdf_testvec hkdf_sha256_tv[] = {
 			"\xd1\x4a\x7e\xe8\x3a\xa0\x57\xa9\x3d\x59\xb0\xa1\x31\x7f\xf0\x9d"
 			"\x10\x5c\xce\xcf\x53\x56\x92\xb1\x4d\xd5",
 		.okm_size = 42,
+		.fips_skip = 1,
 	}, {
 		.test = "unsalted hkdf test with zero info",
 		.ikm  = "\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c"
@@ -244,6 +250,7 @@ static const struct hkdf_testvec hkdf_sha256_tv[] = {
 			"\x50\x04\xbe\x2b\xb9\xbe\x68\xbc\x14\x06\x63\x6f\x54\xbd\x33\x8a"
 			"\x66\xa2\x37\xba\x2a\xcb\xce\xe3\xc9\xa7",
 		.okm_size = 42,
+		.fips_skip = 0,
 	}
 };
 
@@ -265,6 +272,7 @@ static const struct hkdf_testvec hkdf_sha384_tv[] = {
 			"\x06\x3e\x25\xb5\x16\xdc\xbf\x36\x9f\x39\x4c\xfa\xb4\x36\x85\xf7"
 			"\x48\xb6\x45\x77\x63\xe4\xf0\x20\x4f\xc5",
 		.okm_size = 42,
+		.fips_skip = 1,
 	}, {
 		.test = "hkdf test with long input",
 		.ikm  = "\x00\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f"
@@ -296,6 +304,7 @@ static const struct hkdf_testvec hkdf_sha384_tv[] = {
 			"\xf4\xc0\x9c\x55\x00\xdc\xa2\x4a\x56\x22\x2f\xea\x53\xd8\x96\x7a"
 			"\x8b\x2e",
 		.okm_size = 82,
+		.fips_skip = 0,
 	}, {
 		.test = "hkdf test with zero salt and info",
 		.ikm  = "\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b"
@@ -313,6 +322,7 @@ static const struct hkdf_testvec hkdf_sha384_tv[] = {
 			"\x85\x40\x62\xe5\x4c\x73\xa7\xab\xc7\x43\xfa\xde\x9b\x24\x2d\xaa"
 			"\xcc\x1c\xea\x56\x70\x41\x5b\x52\x84\x9c",
 		.okm_size = 42,
+		.fips_skip = 1,
 	}, {
 		.test = "hkdf test with short input",
 		.ikm  = "\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b",
@@ -329,6 +339,7 @@ static const struct hkdf_testvec hkdf_sha384_tv[] = {
 			"\x48\xca\xfe\x53\x24\x9f\x7f\xfe\x14\x97\xa1\x63\x5b\x20\x1f\xf1"
 			"\x85\xb9\x3e\x95\x19\x92\xd8\x58\xf1\x1a",
 		.okm_size = 42,
+		.fips_skip = 1,
 	}, {
 		.test = "unsalted hkdf test with zero info",
 		.ikm  = "\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c"
@@ -348,6 +359,7 @@ static const struct hkdf_testvec hkdf_sha384_tv[] = {
 			"\x7e\x2b\x2c\xd6\xca\x43\x73\xa1\xf3\x14\x50\xa7\xbd\xf9\x48\x2b"
 			"\xfa\xb8\x11\xf5\x54\x20\x0e\xad\x8f\x53",
 		.okm_size = 42,
+		.fips_skip = 0,
 	}
 };
 
@@ -370,6 +382,7 @@ static const struct hkdf_testvec hkdf_sha512_tv[] = {
 			"\xc8\xe2\x6a\x1a\x16\xed\x34\xd9\xfc\x7f\xe9\x2c\x14\x81\x57\x93"
 			"\x38\xda\x36\x2c\xb8\xd9\xf9\x25\xd7\xcb",
 		.okm_size = 42,
+		.fips_skip = 1,
 	}, {
 		.test = "hkdf test with long input",
 		.ikm  = "\x00\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f"
@@ -402,6 +415,7 @@ static const struct hkdf_testvec hkdf_sha512_tv[] = {
 			"\xf6\xa2\x05\x6c\xe3\xaf\x1d\xe4\x4d\x57\x20\x97\xa8\x50\x5d\x9e"
 			"\x7a\x93",
 		.okm_size = 82,
+		.fips_skip = 0,
 	}, {
 		.test = "hkdf test with zero salt and info",
 		.ikm  = "\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b"
@@ -420,6 +434,7 @@ static const struct hkdf_testvec hkdf_sha512_tv[] = {
 			"\x6e\xb1\x79\xdc\x20\x4c\x03\x42\x5c\x97\x0e\x3b\x16\x4b\xf9\x0f"
 			"\xff\x22\xd0\x48\x36\xd0\xe2\x34\x3b\xac",
 		.okm_size = 42,
+		.fips_skip = 1,
 	}, {
 		.test = "hkdf test with short input",
 		.ikm  = "\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b\x0b",
@@ -437,6 +452,7 @@ static const struct hkdf_testvec hkdf_sha512_tv[] = {
 			"\x01\x87\x5d\xb1\xca\x55\xf6\x8c\xfc\xf3\x95\x4d\xc8\xaf\xf5\x35"
 			"\x59\xbd\x5e\x30\x28\xb0\x80\xf7\xc0\x68",
 		.okm_size = 42,
+		.fips_skip = 1,
 	}, {
 		.test = "unsalted hkdf test with zero info",
 		.ikm  = "\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c\x0c"
@@ -458,6 +474,7 @@ static const struct hkdf_testvec hkdf_sha512_tv[] = {
 			"\xb0\xc7\xf6\x3d\x68\xeb\x1a\x80\xea\xf0\x7e\x95\x3c\xfc\x0a\x3a"
 			"\x52\x40\xa1\x55\xd6\xe4\xda\xa9\x65\xbb",
 		.okm_size = 42,
+		.fips_skip = 0,
 	}
 };
 
@@ -467,6 +484,9 @@ static int hkdf_test(const char *shash, const struct hkdf_testvec *tv)
 	unsigned int prk_size;
 	const char *driver;
 	int err;
+	if (fips_enabled && tv->fips_skip) {
+		return 0;
+	}
 
 	tfm = crypto_alloc_shash(shash, 0, 0);
 	if (IS_ERR(tfm)) {
-- 
2.43.0

